### DISCUSSION TITLE
[mini.completion] LSP completion items can be incomplete

### DISCUSSION THREAD

**Discussion by justus2510** [nvim-mini/mini.nvim#1057] [General]
I noticed that the LSP completion is sometimes not listing all items. After some investigation, I believe this is because CompletionList.isIncomplete isn't handled. The [spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_completion) says: 

> for speed clients should be able to filter an already received completion list if the user continues typing. Servers can opt out of this using a [CompletionList](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionList) and mark it as isIncomplete.

This leads to weird situations where typing slow and typing fast leads to different results (because of the delay):
![slow](https://github.com/user-attachments/assets/04089dda-dc88-4938-9795-53331437119d)
![fast](https://github.com/user-attachments/assets/28675d41-16e7-4495-b7d5-8ec7adce13f6)

I prefer to have the delay set to 0, which just makes it worse. The only way around this is to use ctrl+space to force another request.

I'm not sure if this is intended to keep the plugin minimal or not (like the situation with workspace symbols), so I didn't want to open an issue.

**Comment by echasnovski** [comment]
> I'm not sure if this is intended to keep the plugin minimal or not (like the situation with workspace symbols), so I didn't want to open an issue.

Thanks for the thought, I appreciate that!

It is indeed a known behavior which probably won't get resolved. The reason for this is that 'mini.completion' is designed to rely on built-in completion capabilities as much as possible. In particular, this means that after the results are received from the LSP server, they are shown/filtered/unfiltered with built-in mechanics. Which doesn't have the concept of "this *might* be not all completion candidates, please request the rest".

Forcing with `<C-Space>` is the suggested approach here. And overall design is tailored to having not trivial delay, which indeed might be useful here.
